app/modules/sjson.c

```diff
@@ -156,7 +156,7 @@ static void push_number(JSN_DATA *data, struct jsonsl_state_st *state) {
   const char *end = start + state->pos_cur - state->pos_begin;
   lua_pushlstring(data->L, start, end - start);
 #if LUA_VERSION_NUM >= 503
-  int sz = lua_stringtonumber(L, lua_tostring(L, -1));
+  int sz = lua_stringtonumber(data->L, lua_tostring(data->L, -1));
   if (sz) {
     lua_pop(data->L, 1);
```

Comment:
@psjg, You've lost me on this one.  
-   We've changed the throwing behaviour.  With 5.1 anything other than a valid number fails silently pusing 0 onto the stack.  Now we throw an error.  Do we want to do this, or should we continue the existing behaviour?
-  I can't see the added benefit of pushing the numeric string onto the stack and then immediately discarding it, when doing a `lua_stringtonumber(data->L,  start)` does the same but skips this extra stack operation.  Why not just
```C
#if LUA_VERSION_NUM == 501
  lua_pushlstring(data->L, start, end - start);
#else
  if (!lua_stringtonumber(L, start))
    lua_pushinteger(L, 0);
```
which preserves the existing behaviour

Reply:
I'd prefer to fix the 5.1 code to throw if the number cannot be parsed -- it seems to be less unfriendly than silently converting the number to 0. The original code just assumes that the LUA number parser can convert all valid JSON numbers into a value. Thinking about this some more, there are valid JSON numbers that are not representable in LUA -- but I don;'t know whether they cause parsing failures -- e.g. 1e400 which is outside the range of a float or double.

Reply:
I've just checked the behaviour.  You are right.  This uses a `lobject.o` function `luaO_str2num()` where its source states that this follows POSIX `strtod()`.  Stepping though a test case in gdb shows that this is partially incorrect in that `strtod` will stop on any non-valid delimiter, but the  `luaO_str2num()` function will return a conversion error is any tail is found after the end of the number scan.  Bottom line:  pushing the string _is_ the simplest approach.  I withdraw my comment.  Thanks.